File: snippets/fsharp/System/NullReferenceException/Overview/Chain2.fs
module Chain2

type Page() =
    [<DefaultValue>]
    val mutable public URL: System.Uri
    [<DefaultValue>]
    val mutable public Title: string

type Pages() =
    let pages = Array.zeroCreate<Page> 10
    let mutable i = 0

    member _.CurrentPage
        with get () = pages[i]
        and set (value) =
            // Move all the page objects down to accommodate the new one.
            if i > pages.GetUpperBound 0 then
                for ndx = 1 to pages.GetUpperBound 0 do
                    pages[ndx - 1] <- pages[ndx]

            pages[i] <- value
            if i < pages.GetUpperBound 0 then
                i <- i + 1

    member _.PreviousPage =
        if i = 0 then
            if box pages[0] = null then
                Unchecked.defaultof<Page>
            else
                pages[0]
        else
            i <- i - 1
            pages[i + 1]

// <Snippet7>
let pages = Pages()
let current = pages.CurrentPage
if box current <> null then
    let title = current.Title
    printfn $"Current title: '{title}'"
else
    printfn "There is no page information in the cache."
// The example displays the following output:
//       There is no page information in the cache.
// </Snippet7>

File: snippets/csharp/System/NullReferenceException/Overview/Chain2.cs
// <Snippet7>
using System;

public class Chain2Example
{
    public static void Main()
    {
        var pages = new Pages();
        Page current = pages.CurrentPage;
        if (current != null)
        {
            string title = current.Title;
            Console.WriteLine($"Current title: '{title}'");
        }
        else
        {
            Console.WriteLine("There is no page information in the cache.");
        }
    }
}
// The example displays the following output:
//       There is no page information in the cache.
// </Snippet7>

public class Pages
{
    readonly Page[] _page = new Page[10];
    int _ctr = 0;

    public Page CurrentPage
    {
        get { return _page[_ctr]; }
        set
        {
            // Move all the page objects down to accommodate the new one.
            if (_ctr > _page.GetUpperBound(0))
            {
                for (int ndx = 1; ndx <= _page.GetUpperBound(0); ndx++)
                    _page[ndx - 1] = _page[ndx];
            }
            _page[_ctr] = value;
            if (_ctr < _page.GetUpperBound(0))
                _ctr++;
        }
    }

    public Page PreviousPage
    {
        get
        {
            if (_ctr == 0)
            {
                if (_page[0] == null)
                    return null;
                else
                    return _page[0];
            }
            else
            {
                _ctr--;
                return _page[_ctr + 1];
            }
        }
    }
}

public class Page
{
    public Uri URL;
    public string Title;
}
